### ISSUE TITLE
Darknet Yolo v2 added to the OpenCV

### ISSUE THREAD

**Issue by AlexeyAB** [AlexeyAB/darknet#227]
Now Darknet Yolo v2 added to the OpenCV: https://github.com/opencv/opencv/pull/9705
You can use it from master-branch or since OpenCV 3.4.0 will be released.
There is example how to use Yolo v2: https://github.com/opencv/opencv/blob/master/samples/dnn/yolo_object_detection.cpp
For now it CPU-only (~10 times faster than original) and can be used on mobile devices, but in the future GPU will be supported.

If there will be many upvotes, perhaps I will add support for Yolo 9000 :)
**Don't write anything there, just upvote :)**

![yolo](https://user-images.githubusercontent.com/4096485/31386280-ea9db638-adce-11e7-9e07-e56ecd217340.jpg)

**Comment by TheMikeyR** [comment]
Are you planning on making a python wrapper for the OpenCV library? 
Anyway great work!

**Comment by yanchao12122** [comment]
Can you give an example for classification in opencv dnn module？

**Comment by StevenPuttemans** [comment]
> Can you give an example for classification in opencv dnn module？

https://github.com/opencv/opencv/blob/master/samples/dnn/yolo_object_detection.cpp

**Comment by AlexeyAB** [comment]
@yanchao12122 Classification using Yolo in opencv dnn module isn't implemented yet.

So you can use classification:
* Using caffe-model: https://github.com/opencv/opencv/blob/master/samples/dnn/caffe_googlenet.cpp
* Using tensorflow-model: https://github.com/opencv/opencv/blob/master/samples/dnn/tf_inception.cpp

**Comment by AlexeyAB** [comment]
@TheMikeyR It should be generated automatically in the 2 files during installation OpenCV: `pyopencv_generated_funcs.h` and `pyopencv_generated_ns_reg.h`

You can use Yolo in Python in the same way as mobile-SSD-detector, just replace `readNetFromCaffe` to `readNetFromDarknet` and pass yolo-voc.cfg and yolo-voc.weights files: https://github.com/opencv/opencv/blob/master/samples/dnn/mobilenet_ssd_python.py
If there are any errors, please report this to the OpenCV-issues: https://github.com/opencv/opencv/issues

----

1. `/build/modules/python3/`**`pyopencv_generated_funcs.h`**

```
static PyObject* pyopencv_cv_dnn_readNetFromDarknet(PyObject* , PyObject* args, PyObject* kw)
{
    using namespace cv::dnn;

    PyObject* pyobj_cfgFile = NULL;
    String cfgFile;
    PyObject* pyobj_darknetModel = NULL;
    String darknetModel;
    Net retval;

    const char* keywords[] = { "cfgFile", "darknetModel", NULL };
    if( PyArg_ParseTupleAndKeywords(args, kw, "O|O:readNetFromDarknet", (char**)keywords, &pyobj_cfgFile, &pyobj_darknetModel) &&
        pyopencv_to(pyobj_cfgFile, cfgFile, ArgInfo("cfgFile", 0)) &&
        pyopencv_to(pyobj_darknetModel, darknetModel, ArgInfo("darknetModel", 0)) )
    {
        ERRWRAP2(retval = cv::dnn::readNetFromDarknet(cfgFile, darknetModel));
        return pyopencv_from(retval);
    }

    return NULL;
}
```

2. `/build/modules/python3/`**`pyopencv_generated_ns_reg.h`**

```
    {"readNetFromDarknet", (PyCFunction)pyopencv_cv_dnn_readNetFromDarknet, METH_VARARGS | METH_KEYWORDS, "readNetFromDarknet(cfgFile[, darknetModel]) -> retval\n.   @brief Reads a network model stored in <a href=\"https://pjreddie.com/darknet/\">Darknet</a> model files.\n.   *  @param cfgFile      path to the .cfg file with text description of the network architecture.\n.   *  @param darknetModel path to the .weights file with learned network.\n.   *  @returns Network object that ready to do forward, throw an exception in failure cases.\n.   * @details This is shortcut consisting from DarknetImporter and Net::populateNet calls."},
 ```

**Comment by oishi89** [comment]
Hi @AlexeyAB 
Thanks for your amazing work. I'm trying to run the yolo_object_detection.cpp by Xcode but I got the issue "Use of undeclared identifier 'readNetFromDarknet'" and "No member named 'empty' in 'cv::dnn::Net'"

<img width="983" alt="screen shot 2017-10-18 at 11 36 45 am" src="https://user-images.githubusercontent.com/3854757/31736129-bbd60b7c-b3f8-11e7-8b88-014ff9627eb0.png">

 Do you have any ideas ?
Thanks

**Comment by AlexeyAB** [comment]
@oishi89 Hi, what OpenCV version do you use?

It will be possible to use Darknet since OpenCV 3.4.0 that has not yet been released: https://github.com/AlexeyAB/darknet/issues/227#issue-264207848

Also you can use it right now: clone current OpenCV version directly from github using Git, or just download opencv.zip, unpack and build it: https://github.com/opencv/opencv

![image](https://user-images.githubusercontent.com/4096485/31743328-6e4295c2-b463-11e7-8bb1-3bb45df8544c.png)

**Comment by oishi89** [comment]
I have just realized that so I download the source running script to generate opencv2.framework.
https://docs.opencv.org/2.4/doc/tutorials/introduction/ios_install/ios_install.html
But after that I got a number of issues related to C++ 
<img width="1242" alt="screen shot 2017-10-18 at 2 36 29 pm" src="https://user-images.githubusercontent.com/3854757/31743931-c3ee234e-b411-11e7-9fa3-f796c71a471f.png">

**Comment by AlexeyAB** [comment]
@oishi89 You can't build OpenCV and can't build OpenCV examples? Can you build this example https://github.com/opencv/opencv/blob/master/samples/dnn/ssd_object_detection.cpp ?
Try to ask this question here: https://github.com/opencv/opencv/issues

**Comment by oishi89** [comment]
@AlexeyAB I'm trying to run on iOS platform and I think I'm failed at making frame working for iOS

**Comment by oishi89** [comment]
Hi @AlexeyAB 
I resolved the issues above and I was able to build and run the code however I have modified something for iOS platform.
Unfortunately, I got another issue could you take a look and give me some comment. Highly appreciated thanks.
<img width="1244" alt="screen shot 2017-10-18 at 10 14 26 pm" src="https://user-images.githubusercontent.com/3854757/31755289-37e9dfdc-b452-11e7-8d8f-782ce036b664.png">

**Comment by oishi89** [comment]
@AlexeyAB Could you help explain the  cv::Mat detectionMat = net.forward("detection_out");  

<img width="985" alt="screen shot 2017-10-19 at 12 41 23 am" src="https://user-images.githubusercontent.com/3854757/31759444-4b674478-b466-11e7-908b-66a1931cd99f.png">

**Comment by oishi89** [comment]
@AlexeyAB finally I was able to run Yolo on mobile devices. Unfortunately Opencv could not take any advantages of GPU on mobile devices because Opencv only supports CUDA NIDIVA. So the performance on mobile is very bad. I learned that mobilenet backend would be more efficient for mobile so I found out a way to combine Yolo and mobilenet backend as a feature extractor. 
https://github.com/experiencor/basic-yolo-keras
My question is If I trained a yolo model with mobilenet backend following the guideline above. Would It be possible for Opencv dnn ? 
Notice: The training is going to use keras instead of darknet.
Thanks

**Comment by AlexeyAB** [comment]
@oishi89 Yes, OpenCV will support any GPU later by using Halide or/and OpenCL backends.

Now you can try to use `tiny-yolo-voc.weights` + `tiny-yolo-voc.cfg`, and set `width=288` and `height=288` or lower values in the tiny-yolo-voc.cfg file, it will much more faster than full Yolo v2: https://pjreddie.com/darknet/yolo/

You can train your own model based on the tiny-yolo-voc.cfg with the same way as described here: https://github.com/AlexeyAB/darknet#how-to-train-to-detect-your-custom-objects

----

No, you can't use Yolo v2 trained on Keras for OpenCV-dnn module.

Do you talk about **MobileNet**-SSD that supported by OpenCV-dnn module? https://github.com/chuanqi305/MobileNet-SSD/blob/master/MobileNetSSD_deploy.prototxt

What FPS can you get using MobileNet-SSD?

* Keras has backends: **TensorFlow**, Theano, CNTK. But 
* OpenCV-dnn supports DNN Frameworks: Caffe, Darknet, Torch, **Tensorflow**

But OpenCV-dnn Tensorflow Importer doesn't support Region-layer - the main and the last layer of Yolo v2, so you can't use Yolo v2 trained on Keras with OpenCV-dnn.

**Comment by oishi89** [comment]
@AlexeyAB I did trained my own model based on tiny-yolo-voc and I tried it on Opencv as well. The result is around 5 - 6 time faster than full-yolo so it is 0.3 fps for FULL and 1.6 for Tiny. A little bit improvement but it's not fast as my expectation however the accuracy is lower than full. 

For MobileNet-SSD I'm trying that to see how much faster it is. But I got an issue with net.forward(). 
`Requsted layer "" not found in function getLayerData`
I have not resolved it yet. btw if you have any idea please let me know.

This is my demo that using darknet yolo v2 for hand detection. I converted yolo v2 weight to CoreML in this demo but to do that I had to edit yolo-voc.cfg to remove the reorg layer because CoreML tool doesn't support reorg layer. I worried about the accuracy if I removed the layer but the result was very good. Could you please explain why
https://www.youtube.com/watch?v=p3-3kN_fIz0

**Comment by AlexeyAB** [comment]
@oishi89 So did you get yolo-voc.cfg, remove only reorg-layer, then train your own model using https://github.com/experiencor/basic-yolo-keras and then run this model using https://github.com/hollance/YOLO-CoreML-MPSNNGraph ?

CoreML is the fastest way to run Neural Networks on apple devices. Also reorg-layer in Darknet isn't well implemented, so Yolo can be used without reorg-layer (for example as Yolo9000) - some details here: https://github.com/opencv/opencv/pull/9705#discussion_r143230599

So if you can train your model using Keras+Yolo and use it with CoreML without reorg-layer - do it, this is the fastest way.

----

Did you use this C++ example of MobileNet-SSD where called `Mat detection = net.forward("detection_out");`? https://github.com/opencv/opencv/blob/master/samples/dnn/ssd_mobilenet_object_detection.cpp#L108

Did you use these particular model files in particular?
* https://drive.google.com/uc?export=download&id=0B3gersZ2cHIxRm5PMWRoTkdHdHc
* https://raw.githubusercontent.com/chuanqi305/MobileNet-SSD/daef68a6c2f5fbb8c88404266aa28180646d17e0/MobileNetSSD_deploy.prototxt

**Comment by oishi89** [comment]
@AlexeyAB Just want to correct that I get yolo-voc.cfg, remove only reorg-layer, then train your own model using Darknet tool on Window https://timebutt.github.io/static/how-to-train-yolov2-to-detect-custom-objects/. 

Yes CoreML is the fastest way to run Neural Networks on apple devices but my full yolo model's 200MB so I'm dealing with performance problem still. Despise of using GPU the FPS for iPhone 5se : 3.6 and iPhone 7s : 5.5. That why I'm thinking about using Yolo with Mobilenet hopefully it's helpful. I had a quick compare between SSD vs Yolo V2 and in my opinion I prefer Yolo v2.

**Comment by oishi89** [comment]
@AlexeyAB Actually I had to convert from darknet weight file to Keras .h5 then convert to CoreML because CoreML tool only support Keras, caffe, ... except darknet. 
I used those script for conversion from darknet to keras 
https://github.com/allanzelener/YAD2K
Unfortunately, it worked perfect with full yolo v2 but tiny yolo so I have never had a change to run tiny yolo that trained by Darknet on apple devices. 
This is the issue https://github.com/allanzelener/YAD2K/issues/65
If you have time and don't mind could you please take a look. I'm not a professional ML so it's too hard for me.
Thanks in advance.

**Comment by fabito** [comment]
Hi @AlexeyAB ,

I'm trying to load this pre trained model: https://github.com/dannyblueliu/YOLO-version-2-Face-detection/blob/master/cfg/yolo-face.cfg . The following errors are being raised:

```
what():  /opt/opencv-3.4.0-rc/modules/dnn/src/darknet/darknet_io.cpp:503: error: (-212) Unknown layer type: crop in function ReadDarknetFromCfgFile

what():  /opt/opencv-3.4.0-rc/modules/dnn/src/darknet/darknet_io.cpp:503: error: (-212) Unknown layer type: connected in function ReadDarknetFromCfgFile
```
Is it something that still not supported by opencv dnn module or something that can be adjusted in the configuration ?

**Comment by AlexeyAB** [comment]
@fabito Connected and crop layer isn't supported in OpenCV-Darknet yet.
